File: firefox_engine.py
from pathlib import Path

from language_packages import (
    get_installed_pairs,
    get_model_architecture,
    is_package_installed,
    needed_pairs_for_path,
    resolve_model_path,
)
from threaded_engine import ThreadedEngine
from translation_result import TranslationResult
from translation_route import english_pivot_route, translate_with_english_pivot


class FirefoxEngine(ThreadedEngine):
    """Оффлайн-переводчик на моделях Firefox Translations.

    fxtranslate держит нативный указатель и привязан к потоку, в котором создан.
    Поэтому все Translator живут в одном рабочем потоке.
    """

    def __init__(self) -> None:
        super().__init__("firefox-engine")

    def translation_route(self, source_code: str, target_code: str) -> str | None:
        """Возвращает 'direct', 'en' или None, если пути нет."""
        return english_pivot_route(is_package_installed, source_code, target_code)

    def _create_state(self):
        """Кэш Translator по размеру модели и паре языков."""
        return {}

    def _worker_invalidate(self, state) -> None:
        """Сбрасывает загруженные Translator."""
        state.clear()

    def _worker_translate(
        self,
        state,
        text: str,
        source_code: str,
        target_code: str,
    ) -> TranslationResult:
        """Выполняет прямой перевод или двойной через английский."""

        def translate_direct(source_text: str, from_code: str, to_code: str) -> str:
            translator = self._load_translator(state, from_code, to_code)
            return self._run_translate(translator, source_text)

        return translate_with_english_pivot(
            text,
            source_code,
            target_code,
            is_package_installed,
            translate_direct,
            needed_pairs_for_path,
        )

    def _load_translator(
        self,
        translators: dict[tuple[str, str, str], object],
        source_code: str,
        target_code: str,
    ):
        """Читает файлы модели и создаёт Translator в текущем потоке."""
        key = (get_model_architecture(), source_code, target_code)
        cached = translators.get(key)
        if cached is not None:
            return cached

        from fxtranslate import Translator

        model_path = resolve_model_path(source_code, target_code)
        if model_path is None:
            raise FileNotFoundError(
                f"Нет установленной модели {source_code} -> {target_code}"
            )
        model_file = _required_file(model_path, "model.bin")
        src_vocab, trg_vocab = _read_vocabs(model_path)
        shortlist_path = model_path / "lex.bin"
        shortlist = shortlist_path.read_bytes() if shortlist_path.is_file() else None

        translator = Translator(model_file.read_bytes(), src_vocab, trg_vocab, shortlist)
        translators[key] = translator
        return translator

    def _run_translate(self, translator, text: str) -> str:
        """translate_long режет длинный текст на предложения."""
        try:
            translated = translator.translate_long(text)
        except Exception:
            translated = translator.translate(text)
        if not isinstance(translated, str) or not translated.strip():
            raise RuntimeError("Пустой ответ переводчика")
        return translated


def _required_file(model_path: Path, name: str) -> Path:
    """Возвращает обязательный файл модели."""
    file_path = model_path / name
    if not file_path.is_file():
        raise FileNotFoundError(f"В пакете нет файла {name}: {model_path}")
    return file_path


def _read_vocabs(model_path: Path) -> tuple[bytes, bytes]:
    """Читает общий или раздельный словарь SentencePiece."""
    shared = model_path / "vocab.spm"
    if shared.is_file():
        data = shared.read_bytes()
        return data, data
    src_vocab = model_path / "srcvocab.spm"
    trg_vocab = model_path / "trgvocab.spm"
    if src_vocab.is_file() and trg_vocab.is_file():
        return src_vocab.read_bytes(), trg_vocab.read_bytes()
    raise FileNotFoundError(f"В пакете нет словаря: {model_path}")


_engine: FirefoxEngine | None = None


def get_engine() -> FirefoxEngine:
    """Возвращает общий движок приложения."""
    global _engine
    if _engine is None:
        _engine = FirefoxEngine()
    return _engine


def installed_language_codes() -> set[str]:
    """Коды языков из установленных пакетов Firefox."""
    codes: set[str] = set()
    for from_code, to_code, _from_name, _to_name in get_installed_pairs():
        codes.add(from_code)
        codes.add(to_code)
    return codes

File: cpp/src/firefox_engine.cpp
#include "offline_translator/firefox_engine.hpp"

#include "offline_translator/firefox_model_manager.hpp"
#include "offline_translator/route_planner.hpp"

#include <mutex>
#include <stdexcept>
#include <utility>

#ifdef _WIN32
#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <windows.h>
#endif

namespace offline_translator {
namespace {

// Загруженный fxbridge.dll: функции C-ABI моста.
struct FxBridge {
    void* (*engine_create)(
        const char* model,
        const char* src_vocab,
        const char* trg_vocab,
        const char* shortlist) = nullptr;
    char* (*translate)(void* engine, const char* text) = nullptr;
    char* (*translate_long)(void* engine, const char* text) = nullptr;
    void (*string_free)(char* value) = nullptr;
    char* (*backend)() = nullptr;
    void (*engine_destroy)(void* engine) = nullptr;
    char* (*last_error)() = nullptr;

    bool valid() const {
        return engine_create && translate && translate_long &&
               string_free && engine_destroy;
    }
};

FxBridge& bridge() {
    static FxBridge instance = [] {
        FxBridge loaded{};
#ifdef _WIN32
        HMODULE module = ::LoadLibraryW(L"fxbridge.dll");
        if (!module) {
            return loaded;
        }
        auto resolve = [module](const char* name) {
            return reinterpret_cast<void*>(::GetProcAddress(module, name));
        };
        loaded.engine_create =
            reinterpret_cast<void* (*)(
                const char*, const char*, const char*, const char*)>(
                resolve("fxt_engine_create"));
        loaded.translate =
            reinterpret_cast<char* (*)(void*, const char*)>(
                resolve("fxt_translate"));
        loaded.translate_long =
            reinterpret_cast<char* (*)(void*, const char*)>(
                resolve("fxt_translate_long"));
        loaded.string_free =
            reinterpret_cast<void (*)(char*)>(resolve("fxt_string_free"));
        loaded.backend =
            reinterpret_cast<char* (*)()>(resolve("fxt_backend"));
        loaded.last_error =
            reinterpret_cast<char* (*)()>(resolve("fxt_last_error"));
        loaded.engine_destroy =
            reinterpret_cast<void (*)(void*)>(resolve("fxt_engine_destroy"));
#endif
        return loaded;
    }();
    return instance;
}

const FxBridge& require_bridge() {
    const auto& loaded = bridge();
    if (!loaded.valid()) {
        throw std::runtime_error(
            "Не найден fxbridge.dll (движок Firefox Translations). "
            "Переустановите программу.");
    }
    return loaded;
}

std::string read_bridge_string(char* (*function)()) {
    if (!function) {
        return {};
    }
    char* value = function();
    if (!value) {
        return {};
    }
    std::string result(value);
    require_bridge().string_free(value);
    return result;
}

// Диагностика из моста: последняя ошибка на текущем потоке.
std::string bridge_last_error() {
    const auto& loaded = bridge();
    if (!loaded.last_error) {
        return {};
    }
    char* value = loaded.last_error();
    if (!value) {
        return {};
    }
    std::string result(value);
    loaded.string_free(value);
    return result;
}

}  // namespace

struct FirefoxEngine::PairState {
    void* engine{nullptr};
    std::string backend;

    PairState(
        const std::filesystem::path& models_root,
        const std::string& architecture,
        const std::string& source_code,
        const std::string& target_code) {
        const auto& loaded = require_bridge();
        FirefoxModelManager manager(
            models_root, architecture, source_code, target_code);
        const auto model_path = manager.resolve_model_path();
        if (!model_path) {
            throw std::runtime_error(
                "Нет установленной модели " + source_code + " -> " +
                target_code);
        }
        auto vocab_path = *model_path / "vocab.spm";
        if (!std::filesystem::is_regular_file(vocab_path)) {
            vocab_path = *model_path / "srcvocab.spm";
        }
        if (!std::filesystem::is_regular_file(vocab_path)) {
            throw std::runtime_error(
                "В пакете нет словаря: " + model_path->string());
        }
        auto trg_vocab_path = *model_path / "vocab.spm";
        if (!std::filesystem::is_regular_file(trg_vocab_path)) {
            trg_vocab_path = *model_path / "trgvocab.spm";
        }
        const auto shortlist_path = *model_path / "lex.bin";
        const bool has_shortlist =
            std::filesystem::is_regular_file(shortlist_path);
        const auto model_file = *model_path / "model.bin";
        if (!std::filesystem::is_regular_file(model_file)) {
            throw std::runtime_error(
                "В пакете нет файла модели: " + model_file.string());
        }
        engine = loaded.engine_create(
            model_file.string().c_str(),
            vocab_path.string().c_str(),
            trg_vocab_path.string().c_str(),
            has_shortlist ? shortlist_path.string().c_str() : nullptr);
        if (!engine) {
            const auto details = bridge_last_error();
            throw std::runtime_error(
                "Не удалось создать движок Firefox для " + source_code +
                " -> " + target_code +
                (details.empty() ? std::string{} : ": " + details));
        }
        backend = read_bridge_string(loaded.backend);
    }

    ~PairState() {
        if (engine) {
            // Мост может быть уже выгружен при завершении процесса —
            // проверяем указатель перед вызовом.
            if (bridge().valid()) {
                bridge().engine_destroy(engine);
            }
            engine = nullptr;
        }
    }
};

FirefoxEngine::FirefoxEngine(
    std::filesystem::path models_root,
    std::string architecture)
    : models_root_(std::move(models_root)),
      architecture_(std::move(architecture)) {
    if (!FirefoxModelManager::is_architecture(architecture_)) {
        throw std::invalid_argument("Неизвестный размер модели Firefox");
    }
}

FirefoxEngine::~FirefoxEngine() {
    stop();
}

TranslationResult FirefoxEngine::translate(
    std::string_view text,
    std::string_view source_code,
    std::string_view target_code) {
    if (stopped_) {
        throw std::runtime_error("Движок Firefox остановлен");
    }
    return translate_with_english_pivot(
        text,
        source_code,
        target_code,
        [this](std::string_view source, std::string_view target) {
            return is_installed(source, target);
        },
        [this](
            std::string_view direct_text,
            std::string_view direct_source,
            std::string_view direct_target) {
            return translate_direct(direct_text, direct_source, direct_target);
        },
        "Firefox");
}

void FirefoxEngine::warmup(
    std::string_view source_code,
    std::string_view target_code) {
    static_cast<void>(translate("Hello", source_code, target_code));
}

void FirefoxEngine::invalidate() {
    if (!stopped_) {
        pairs_.clear();
    }
}

void FirefoxEngine::stop() {
    if (!stopped_) {
        stopped_ = true;
        pairs_.clear();
    }
}

std::optional<std::string> FirefoxEngine::translation_route(
    std::string_view source_code,
    std::string_view target_code) const {
    return english_pivot_route(
        [this](std::string_view source, std::string_view target) {
            return is_installed(source, target);
        },
        source_code,
        target_code);
}

bool FirefoxEngine::is_installed(
    std::string_view source_code,
    std::string_view target_code) const {
    return FirefoxModelManager(
               models_root_,
               architecture_,
               std::string(source_code),
               std::string(target_code))
        .is_installed();
}

std::string FirefoxEngine::translate_direct(
    std::string_view text,
    std::string_view source_code,
    std::string_view target_code) {
    auto& pair_state = get_pair_state(source_code, target_code);
    const auto& loaded = require_bridge();
    char* raw = loaded.translate_long(
        pair_state.engine,
        std::string(text).c_str());
    if (!raw) {
        throw std::runtime_error("Движок Firefox вернул пустой ответ");
    }
    std::string translated(raw);
    loaded.string_free(raw);
    while (!translated.empty() &&
           (translated.back() == ' ' || translated.back() == '\n')) {
        translated.pop_back();
    }
    if (translated.empty()) {
        throw std::runtime_error("Пустой ответ переводчика");
    }
    return translated;
}

FirefoxEngine::PairState& FirefoxEngine::get_pair_state(
    std::string_view source_code,
    std::string_view target_code) {
    const LanguagePair key{
        std::string(source_code),
        std::string(target_code),
    };
    const auto found = pairs_.find(key);
    if (found != pairs_.end()) {
        return *found->second;
    }
    auto state = std::make_unique<PairState>(
        models_root_, architecture_, key.first, key.second);
    auto* state_ptr = state.get();
    pairs_.emplace(key, std::move(state));
    return *state_ptr;
}

}  // namespace offline_translator
